components/script/dom/transformstream.rs

```diff
@@ -0,0 +1,1015 @@
+            // Otherwise:
+            // Perform ! ReadableStreamDefaultControllerError(readable.[[controller]], reason).
+            self.readable.get_default_controller().error(v, can_gc);
+
+            // Resolve controller.[[finishPromise]] with undefined.
+            finish_promise.resolve_native(&(), can_gc);
+        }
+    }
+}
+
+impl js::gc::Rootable for CancelPromiseRejection {}
+
+/// Reacting to rejection of cancelpromise as part of
+/// <https://streams.spec.whatwg.org/#transform-stream-default-sink-abort-algorithm>
+#[derive(JSTraceable, MallocSizeOf)]
+#[cfg_attr(crown, crown::unrooted_must_root_lint::must_root)]
+struct CancelPromiseRejection {
+    readable: Dom<ReadableStream>,
+    controller: Dom<TransformStreamDefaultController>,
+}
+
+impl Callback for CancelPromiseRejection {
+    /// Reacting to backpressureChangePromise with the following fulfillment steps:
+    #[allow(unsafe_code)]
+    fn callback(&self, cx: SafeJSContext, v: SafeHandleValue, _realm: InRealm, can_gc: CanGc) {
+        // Perform ! ReadableStreamDefaultControllerError(readable.[[controller]], r).
+        self.readable.get_default_controller().error(v, can_gc);
+
+        // Reject controller.[[finishPromise]] with r.
+        self.controller
+            .get_finish_promise()
+            .expect("finish promise is not set")
+            .reject(cx, v, can_gc);
+    }
+}
+
+//----------
+impl js::gc::Rootable for SourceCancelPromiseFulfillment {}
+
+/// Reacting to fulfillment of the cancelpromise as part of
+/// <https://streams.spec.whatwg.org/#transform-stream-default-source-cancel>
+#[derive(JSTraceable, MallocSizeOf)]
+#[cfg_attr(crown, crown::unrooted_must_root_lint::must_root)]
+struct SourceCancelPromiseFulfillment {
+    writeable: Dom<WritableStream>,
+    controller: Dom<TransformStreamDefaultController>,
+    stream: Dom<TransformStream>,
+}
+
+impl Callback for SourceCancelPromiseFulfillment {
+    /// Reacting to backpressureChangePromise with the following fulfillment steps:
+    fn callback(&self, cx: SafeJSContext, v: SafeHandleValue, _realm: InRealm, can_gc: CanGc) {
+        // If cancelPromise was fulfilled, then:
+        let finish_promise = self
+            .controller
+            .get_finish_promise()
+            .expect("finish promise is not set");
+
+        let global = &self.writeable.global();
+        // If writable.[[state]] is "errored", reject controller.[[finishPromise]] with writable.[[storedError]].
+        if self.writeable.is_errored() {
+            rooted!(in(*cx) let mut error = UndefinedValue());
+            self.writeable.get_stored_error(error.handle_mut());
+            finish_promise.reject(cx, error.handle(), can_gc);
+        } else {
+            // Otherwise:
+            // Perform ! WritableStreamDefaultControllerErrorIfNeeded(writable.[[controller]], reason).
+            self.writeable
+                .get_default_controller()
+                .error_if_needed(cx, v, global, can_gc);
```

Comment:
`promise_rejects_exactly: closed should reject function "function() { throw e; }" threw undefined but we expected it to throw object "error1: error1 message"`

Here and in the rejection steps, you need to pass the original reason, which isn't `v`, but that passed to `transform_stream_default_source_cancel`

Reply:
Right, so in the rejection steps it should be `v`, but `reason` in the fulfillment ones...